### PR TITLE
Reduce JoinNestedLoop compile time

### PR DESCRIPTION
**Pull request by mrzzzrm**
PR currently still includes #1541, #1541 should be merged first.

Similarly to the ColumnVsColumnScan, the JNL has a dedicated SLOW PATH where it aggressively erases the SegmentTypes (Left+Right) and the PredicateCondition if the DataTypes or the SegmentTypes of the join columns are not the same.

Compile time of libhyrise on pella, single-threaded, release: 54m9,863s (from 57m47,161s on master)
Compile time of join_nested_loop.cpp with the same settings: 57.212s (from 273.552s on master)

### CODE REVIEW

**Comment by mrks** [comment]
could you merge the master so that we have a proper diff, please?

**Comment by mrzzzrm** [comment]
Done

**Comment by Bouncner** [comment]
Won't take a look at it until Monday. :-/

File: src/lib/operators/join_nested_loop.cpp [hunk 1]
```diff
@@ -172,20 +176,62 @@ std::shared_ptr<const Table> JoinNestedLoop::_on_execute() {
...
-                                                JoinNestedLoop::JoinParams& params) {
+void JoinNestedLoop::_join_two_untyped_segments(const BaseSegment& base_segment_left,
+                                                const BaseSegment& base_segment_right, const ChunkID chunk_id_left,
+                                                const ChunkID chunk_id_right, JoinNestedLoop::JoinParams& params) {
   /**
-   * The nested loops.
+   * This function dispatches `join_two_typed_segments`.
+   *
+   * To reduce compile time, we erase the types of Segments and the PredicateCondition/comparator if
+   * `base_segment_left.data_type() !=  base_segment_left.data_type()` or `LeftSegmentType != RightSegmentType`. This is
```
Comment (mrks) [inline src/lib/operators/join_nested_loop.cpp]: format whitespace

File: src/lib/operators/join_nested_loop.cpp [hunk 2]
```diff
@@ -172,20 +176,62 @@ std::shared_ptr<const Table> JoinNestedLoop::_on_execute() {
...
+   * the "SLOW PATH".
+   * If data types and segment types are the same, we take the "FAST PATH", were only the SegmentType of left segment is
```
Comment (mrks) [inline src/lib/operators/join_nested_loop.cpp]: where

File: src/lib/operators/join_nested_loop.cpp [hunk 3]
```diff
@@ -33,10 +32,16 @@ void process_match(RowID left_row_id, RowID right_row_id, const JoinNestedLoop::
 }
 
 // inner join loop that joins two segments via their iterators
+// __attribute__((noinline)) to reduce compile time. As the hotloop is within this function, no expected perf loss
```
[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/operators/join_nested_loop.cpp]: The nested loops are IN this function so I wouldn't expect any difference. If you really want me to, I can microbenchmark this, though

[the remark this answers is not in the document]
Comment (mrks) [reply src/lib/operators/join_nested_loop.cpp]: Just Phrasing.

[the remark this answers is not in the document]
Comment (mrzzzrm) [reply src/lib/operators/join_nested_loop.cpp]: ah

**Pull request merged**